### ISSUE TITLE
Issue with copying tensor with `requires_grad` set to true

### ISSUE THREAD

**Issue by jatkinson1000** [Cambridge-ICCS/FTorch#327]
I had some unexpected behaviour using the overloaded assignment (`=`) operator to set a tensor equal to another that had `requires_grad=.true.`.

It resulted in the following stacktrace (reduced output):
```
libc++abi: terminating due to uncaught exception of type c10::Error: a leaf Variable that requires grad is being used in an in-place operation.

...

frame #4: at::_ops::copy_::call(at::Tensor&, at::Tensor const&, bool) + 304 (0x11124c9dc in libtorch_cpu.dylib)
frame #5: torch_tensor_assign + 92 (0x104756100 in libftorch.dylib)
```

It seems perhaps we cannot copy a tensor that has `requires_grad` set?
Happy to try and rewrite a MWE if needed.

Perhaps there are different versions of "copying" as we sometimes see in python?

From what I read leaf tensors cannot be modified as they break the backprop cycle.
It seems there is a need to use `tensor.clone()` rather than `tensor.copy()`.
Though there are some other hacky suggestions out there like creating a new tensor through an operation:
```
tensor = tensor + 0;
tensor.copy_(source_tensor);
```

**Comment by jwallwork23** [comment]
Oh interesting. We were previously using `detach` and `clone` but I switched to just copying the values in C++ because I didn't think they were needed. Then in #303 we switched to using `std::move`. Maybe we need to explicitly implement cloning.
